### ISSUE TITLE
Gradle doesn't find a usable Java version

### ISSUE THREAD

**Issue by th** [oss-review-toolkit/ort#9293]
### Describe the bug

I'm on Debian bookworm with the Temurin JDK package, version 23 installed.

th@thinkpad ~> echo $JAVA_HOME
/usr/lib/jvm/default-java/

th@thinkpad ~> java -version
openjdk version "23" 2024-09-17
OpenJDK Runtime Environment Temurin-23+37 (build 23+37)
OpenJDK 64-Bit Server VM Temurin-23+37 (build 23+37, mixed mode, sharing)

### To Reproduce

1. ./gradlew installDist

### Expected behavior

ort-tools are built.

### Console / log output

Starting a Gradle Daemon (subsequent builds will be faster)
Daemon JVM discovery is an incubating feature.

FAILURE: Build failed with an exception.

* What went wrong:
Cannot find a Java installation on your machine (Linux 6.1.0-26-amd64 amd64) matching: Compatible with Java 21, vendor matching('Adoptium') (from gradle/gradle-daemon-jvm.properties).

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

### Environment

*Or* manually specify:

* ORT version: 35
* Java version: Temurin 23
* OS: Debian bookworm

### Additional context

Up until ort  34 it built fine. I just checked that again after a git checkout 34.0.0.

There have apparently been some changes about detecting or requiring specific compiler versions (the commits reference "Adoptium", although the Java distribution has been renamed to "Temurin" some time ago).

I tried to set things up using gradle property files, but I'm not versed in Gradle and couldn't find ort-specific documentation how to choose the compiler. I deleted those property file attempts since and unfortunately cannot show them anymore.

**Comment by sschuberth** [comment]
> the commits reference "Adoptium", although the Java distribution has been renamed to "Temurin" some time ago

Yes, [Gradle calls it](https://docs.gradle.org/current/kotlin-dsl/gradle/org.gradle.jvm.toolchain/-jvm-vendor-spec/index.html) still "Adoptium"; there is no "Temurin" enum entry.

**Comment by sschuberth** [comment]
> Cannot find a Java installation on your machine (Linux 6.1.0-26-amd64 amd64) matching: Compatible with Java 21, vendor matching('Adoptium') (from gradle/gradle-daemon-jvm.properties).

The problem is that Java toolchain bootstrapping does not yet work for the Gradle *daemon*. As a work-around, removing this line should do the tick:

https://github.com/oss-review-toolkit/ort/blob/e5c6e0cf4ef9b08dcf1229983fd754697f3a7bca/gradle/gradle-daemon-jvm.properties#L4

After the build succeeds once, you can restore the line again.

**Comment by th** [comment]
Is there some cleanup step I need to do? It doesn't seem to work for me.

toolchainVersion = 21 is just the minimum version, correct?


Starting a Gradle Daemon (subsequent builds will be faster)
Daemon JVM discovery is an incubating feature.

FAILURE: Build failed with an exception.

* What went wrong:
Cannot find a Java installation on your machine (Linux 6.1.0-26-amd64 amd64) matching: Compatible with Java 21, any vendor (from gradle/gradle-daemon-jvm.properties).

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

**Comment by sschuberth** [comment]
> toolchainVersion = 21 is just the minimum version, correct?

Yes.

> Cannot find a Java installation on your machine (Linux 6.1.0-26-amd64 amd64) matching: Compatible with Java 21, any vendor (from gradle/gradle-daemon-jvm.properties).

Then you're probably running into a bug with Gradle's Java toolchain discovery mechanism... running on Java 23 [should](https://github.com/gradle/gradle/issues/29015) be supported as of Gradle 8.10.

**Comment by th** [comment]
Indeed. I just installed Temurin 21, didn't even update JAVA_HOME, and it builds. Fascinating.

Is there a special reason why you limit JDKs to Adoptium/Temurin, some incompatibility, or is it just to cut down on support issues and standardize on one Java distribution?

**Comment by sschuberth** [comment]
> Indeed. I just installed Temurin 21, didn't even update JAVA_HOME, and it builds. Fascinating.

I'd appreciate if you could take the time to report this to https://github.com/gradle/gradle/issues.

> Is there a special reason why you limit JDKs to Adoptium/Temurin

To work around another bug with Gradle's Java toolchain discovery mechanism 😬 See the commit message for https://github.com/oss-review-toolkit/ort/commit/c9d114c58dcfc2620278fc98d7e73059367e3811.

**Comment by sschuberth** [comment]
@th, would you agree that it make sense to close the issue on our side as there's probably nothing we can do about it, and you've filed an upstream issue?
